// ng2-wizard/src/wizard-step.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var wizard_step_tab_component_1 = require("./wizard-step-tab.component");
/**
 * # WizardStepComponent
 *
 * Childrens component of WizardComponent
 *
 *```
 * <wizard-step>
 *  <wizard-step-tab>Title of your tab</wizard-step-tab>
 *  Hello World - your content
 * </wizard-step>
 *```
 *
 * ## Atention!!
 * The tag &lt;wizard-step-tab&gt; is o obrigatory
 *```
 *  <wizard-step-tab>Title of your tab</wizard-step-tab>
 *```
 */
var WizardStepComponent = (function () {
    function WizardStepComponent(elementRef) {
        this.elementRef = elementRef;
        this.isActive = false;
    }
    WizardStepComponent.prototype.ngAfterContentInit = function () {
        if (this.wizardStepTabComponent) {
            this.wizardStepTabComponent.setDisplay('none');
            this.tabName = this.wizardStepTabComponent.getHTML();
        }
        else
            throw new Error('WizardStepComponent: TagName "wizard-step-tab" was not found in step');
    };
    return WizardStepComponent;
}());
__decorate([
    core_1.ContentChild(wizard_step_tab_component_1.WizardStepTabComponent),
    __metadata("design:type", wizard_step_tab_component_1.WizardStepTabComponent)
], WizardStepComponent.prototype, "wizardStepTabComponent", void 0);
WizardStepComponent = __decorate([
    core_1.Component({
        selector: 'wizard-step',
        styles: ["\n      .step {\n          padding: 15px;\n          padding-top: 20px;\n      }\n    "],
        template: "\n      <div class=\"step\" [hidden]=\"!isActive\">\n      \t\t<ng-content></ng-content>\n      </div>\n    "
    }),
    __metadata("design:paramtypes", [core_1.ElementRef])
], WizardStepComponent);
exports.WizardStepComponent = WizardStepComponent;
